### PR TITLE
Integrate latest @fluid-experimental/tree from Whiteboard

### PR DESCRIPTION
**Pull request by Abe27342**
Brings over the latest changes from Whiteboard, including support for a new write format with compression. If consumers have any difficulty migrating to use breaking code changes, please reach out: the history of changes on the Whiteboard end has been preserved (but won't be on main once squashed).

Since there has generally been very little development in the fluidframework repository on this DDS, this PR took the following approach to integrating code:

1. Split a new copy of the tree DDS history from Whiteboard
3. Start from the commit before the most recent integration into fluidframework
4. Delete the contents of experimental/dds/tree (src, docs folders) and re-add them preserving history on the forked branch using `git subtree`
5. Merge all changes made in the fluidframework repository in the last 10 months, resolving conflicts.
6. Fix build/lint errors.

The conflicts resolved in 5 ended up being relatively trivial, one can see this by examining the `git log` from the framework point of view since 6/22/2021.

One sadness with fixing build/lint errors are that because fluidframework uses typescript 4.1, we lose some type safety as this version of typescript doesn't have as much intelligence when it comes to type narrowing. I've left TODOs in all such places mentioning the casts can (and should) be removed on upgrade to typscript 4.6

It's suggested that the reviewer verifies the split git history generated in (1), then reviews by subsequent commit rather than review the PR in toto.

### CODE REVIEW

File: experimental/dds/tree/docs/Breaking-Change-Migration.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
+# Breaking Change Migration
```
Comment (christiango) [inline experimental/dds/tree/docs/Breaking-Change-Migration.md:1]: Are there any docs on the breaking API changes for teams using the current tree from FF? I suspect having the API changes documented will make migrations easier for partners.

**Review by noencke — APPROVED** [review]

File: experimental/dds/tree/docs/Breaking-Change-Migration.md (hunk 1 above)
Comment (taylorsw04) [reply experimental/dds/tree/docs/Breaking-Change-Migration.md:1]: Normally, yes, I agree. However, given timing/deadlines, we took a different approach: there's a small set of consumers (this is marked experimental, after all) and I've worked with each of them to integrate this code already. We could probably do a fast-follow with some documentation on migrating from 0.0.2 to 0.1.1 APIs, but in general I don't think anyone else is using it yet.

Comment (taylorsw04) [reply experimental/dds/tree/docs/Breaking-Change-Migration.md:1]: I'm going to resolve this to unblock this before the branch cut, but we can follow up about supporting it :)

Comment (Abe27342) [reply experimental/dds/tree/docs/Breaking-Change-Migration.md:1]: I added a note to the top of the PR mentioning in theory partner teams should be able to scrub the history to find whatever breaking change is affecting them (and if not, to reach out to me as viewing the history on the Whiteboard end might make this clearer). But yeah, in practice we think all teams have already been notified and in many cases have working prototypes against these changes.

**Pull request merged**